swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,349 @@
+use crate::handler::{InboundUpgradeSend, OutboundUpgradeSend};
+use crate::{
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    NegotiatedSubstream, SubstreamProtocol,
+};
+use futures::future::BoxFuture;
+use futures::future::FutureExt;
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use libp2p_core::upgrade::{NegotiationError, ReadyUpgrade};
+use libp2p_core::UpgradeError;
+use std::collections::VecDeque;
+use std::error::Error;
+use std::fmt;
+use std::future::Future;
+use std::task::{Context, Poll, Waker};
+use void::Void;
+
+/// A low-level building block for protocols that can be expressed as async functions.
+///
+/// An async function in Rust is executed within an executor and thus only has access to its local state
+/// and by extension, the state that was available when the [`Future`] was constructed.
+///
+/// This [`ConnectionHandler`] aims to reduce the boilerplate for protocols which can be expressed as
+/// a static sequence of reads and writes to a socket where the response can be generated with limited
+/// "local" knowledge or in other words, the local state within the [`Future`].
+///
+/// For outbound substreams, arbitrary data can be supplied via [`InEvent::NewOutbound`] which will be
+/// made available to the callback once the stream is fully negotiated.
+///
+/// Inbound substreams may be opened at any time by the remote. To facilitate this one and more usecases,
+/// the supplied callbacks for inbound and outbound substream are given access to the handler's `state`
+/// field. This `state` field can contain arbitrary data and can be updated by the [`NetworkBehaviour`]
+/// via [`InEvent::UpdateState`].
+///
+/// The design of this [`ConnectionHandler`] trades boilerplate (you don't have to write your own handler)
+/// and simplicity (small API surface) for eventual consistency, depending on your protocol design:
+///
+/// Most likely, the [`NetworkBehaviour`] is the authoritive source of `TState` but updates to it have
+/// to be manually performed via [`InEvent::UpdateState`]. Thus, the state given to newly created
+/// substreams, may be outdated and only eventually-consistent.
+///
+/// [`NetworkBehaviour`]: crate::NetworkBehaviour
+pub fn from_fn<TInbound, TOutbound, TOutboundOpenInfo, TState, TInboundFuture, TOutboundFuture>(
+    protocol: &'static str,
+    inbound_streams_limit: usize,
+    pending_dial_limit: usize,
+    on_new_inbound: impl Fn(NegotiatedSubstream, &mut TState) -> TInboundFuture + Send + 'static,
+    on_new_outbound: impl Fn(NegotiatedSubstream, &mut TState, TOutboundOpenInfo) -> TOutboundFuture
+        + Send
+        + 'static,
+) -> FromFn<TInbound, TOutbound, TOutboundOpenInfo, TState>
+where
+    TInboundFuture: Future<Output = TInbound> + Send + 'static,
+    TOutboundFuture: Future<Output = TOutbound> + Send + 'static,
+    TState: Default,
```

Comment:
What do you think about having a `TState` passed as parameter instead of using `Default`? Use case would be config parameters like timeouts, or for example the next_request_id `Arc` in the request-response protocol:
https://github.com/libp2p/rust-libp2p/blob/89f898c69f6e58944e746565c3f8f8a3bbf3324e/protocols/request-response/src/handler.rs#L95-L96

Reply:
Fair enough. I wanted to reduce the number of parameters but having this configurable seems reasonable!

Reply:
Wouldn't function pointers be unnecessarily limiting? What if the user wants to close over certain variables?

Is your argument that they should capture it in `TState` instead?

Reply:
> Is your argument that they should capture it in TState instead?

Yes exactly (provided that we decide to make `TState` a parameter). I don't think we would limit the user with this in any case, and we could reduce complexity a bit. Can you think of any case that would not be possible with this?

Reply:
I think everything would still be possible but it is just not as ergonomic. I'd assume that users are used to closures and having to think how to "emulate" them creates complexity.

Why would it be simpler to have a function pointer?

Reply:
To be clear, I agree making `TState` a parameter but I'd keep the closures.

Reply:
> lexity.
> 
> Why would it be simpler to have a function pointer?

We would not need to box them and it would improve readability a bit. But I also see your point, so I am fine with sticking with the generic `Fn`.

Reply:
Thanks for the suggestion! The boxing is internal so I don't think it concerns users much. I think the best of both worlds is allowing users to pass in `TState` but also allow closures.

Now that I think of it, we can actually make them `FnMut`s to allow even more use cases.